File: modules/void-ssh/ios/VoidSshModule.swift
import ExpoModulesCore
import NMSSH
import CryptoKit
import Foundation

private let keyType = "ssh-ed25519"

enum VoidSshError: Error {
  case unknownConnection
  case connectFailed
  case hostKeyNotVerified
  case authFailed
  case shellFailed
}

private final class LiveConnection {
  let session: NMSSHSession
  var hostKeyConfirmed = false
  var closed = false
  var readThread: Thread?
  init(session: NMSSHSession) { self.session = session }
}

/**
 * Real interactive SSH shell over NMSSH (libssh2) — see
 * modules/void-ssh/README.md. Written carefully against NMSSH's documented
 * API but NOT yet compiled or run on a device (no Xcode toolchain in the
 * environment this was authored in). First step on a real Mac: open the
 * generated `ios/` project after `npx expo prebuild`, let CocoaPods pull
 * NMSSH in, and fix up any NMSSHChannel method-name mismatches Xcode
 * flags — those are the one part of this file most likely to need a
 * small correction against the exact NMSSH version that resolves.
 *
 * Known limitation vs. Android: NMSSHChannel's delegate hands back
 * already-UTF8-decoded NSStrings, not raw bytes, so a multi-byte
 * character split exactly across two reads could in rare cases render as
 * a replacement character. Acceptable for a v1 — flagged here rather than
 * silently claimed as byte-perfect.
 */
public class VoidSshModule: Module, NMSSHChannelDelegate {
  private var connections: [String: LiveConnection] = [:]
  private let lock = NSLock()

  public func definition() -> ModuleDefinition {
    Name("VoidSsh")
    Events("onData", "onStatus")

    AsyncFunction("connect") { (host: String, port: Int, username: String) -> [String: String] in
      guard let session = NMSSHSession(host: "\(host):\(port)", andUsername: username) else {
        throw VoidSshError.connectFailed
      }
      session.connect()
      guard session.isConnected else {
        throw VoidSshError.connectFailed
      }

      let id = UUID().uuidString
      var typeOut: NSString = ""
      let fingerprint = VoidSshHostKey.fingerprint(for: session, type: &typeOut) ?? ""

      self.lock.lock()
      self.connections[id] = LiveConnection(session: session)
      self.lock.unlock()

      self.sendEvent("onStatus", ["connectionId": id, "status": "connected"])
      return ["connectionId": id, "hostKeyFingerprintSha256": fingerprint, "hostKeyType": typeOut as String]
    }

    AsyncFunction("verifyHostKey") { (connectionId: String, accepted: Bool) in
      guard let conn = self.connections[connectionId] else { throw VoidSshError.unknownConnection }
      if accepted {
        conn.hostKeyConfirmed = true
      } else {
        self.closeConnection(connectionId, reason: "host key rejected")
      }
    }

    AsyncFunction("authenticatePassword") { (connectionId: String, username: String, password: String) in
      let conn = try self.requireConfirmed(connectionId)
      conn.session.authenticate(byPassword: password)
      guard conn.session.isAuthorized else { throw VoidSshError.authFailed }
      self.sendEvent("onStatus", ["connectionId": connectionId, "status": "authenticated"])
    }

    AsyncFunction("authenticateKey") { (connectionId: String, username: String, privateKeyOpenSsh: String, passphrase: String?) in
      let conn = try self.requireConfirmed(connectionId)
      // NMSSH's public-key auth takes file paths, not in-memory PEM — write
      // to a NSTemporaryDirectory() file for the duration of the call and
      // remove it immediately after, since the Keychain (not disk) is our
      // real secret store.
      let tmpURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(UUID().uuidString)
      try privateKeyOpenSsh.write(to: tmpURL, atomically: true, encoding: .utf8)
      defer { try? FileManager.default.removeItem(at: tmpURL) }

      conn.session.authenticate(byInMemoryPublicKey: nil, privateKey: tmpURL.path, andPassword: passphrase)
      guard conn.session.isAuthorized else { throw VoidSshError.authFailed }
      self.sendEvent("onStatus", ["connectionId": connectionId, "status": "authenticated"])
    }

    AsyncFunction("startShell") { (connectionId: String, cols: Int, rows: Int) in
      guard let conn = self.connections[connectionId] else { throw VoidSshError.unknownConnection }
      let channel = conn.session.channel
      channel.delegate = self
      channel.requestPty = true
      channel.ptyTerminalType = NMSSHChannelPtyTerminal.xterm

      var startError: NSError?
      channel.startShell(&startError)
      if let startError { throw startError }

      channel.requestSizeWidth(UInt(cols), height: UInt(rows))

      self.sendEvent("onStatus", ["connectionId": connectionId, "status": "shellReady"])
    }

    Function("write") { (connectionId: String, data: String) in
      guard let conn = self.connections[connectionId] else { return }
      var error: NSError?
      conn.session.channel.write(data, error: &error, timeout: 0)
    }

    Function("resize") { (connectionId: String, cols: Int, rows: Int) in
      guard let conn = self.connections[connectionId] else { return }
      conn.session.channel.requestSizeWidth(UInt(cols), height: UInt(rows))
    }

    Function("disconnect") { (connectionId: String) in
      self.closeConnection(connectionId, reason: "closed by app")
    }

    AsyncFunction("generateEd25519KeyPair") { (comment: String) -> [String: String] in
      let priv = Curve25519.Signing.PrivateKey()
      let seed = [UInt8](priv.rawRepresentation)
      let pub = [UInt8](priv.publicKey.rawRepresentation)

      let publicBlob = Self.encodeSshEd25519PublicKey(pub)
      let privateKeyOpenSsh = Self.encodeOpenSshEd25519PrivateKey(seed: seed, pub: pub, comment: comment)
      let fingerprint = "SHA256:" + Self.base64NoPad(Data(SHA256.hash(data: Data(publicBlob))))
      let publicKeySsh = "\(keyType) \(Data(publicBlob).base64EncodedString()) \(comment)"

      return [
        "privateKeyOpenSsh": privateKeyOpenSsh,
        "publicKeySsh": publicKeySsh,
        "fingerprintSha256": fingerprint,
      ]
    }

    OnDestroy {
      for id in Array(self.connections.keys) {
        self.closeConnection(id, reason: "module destroyed")
      }
    }
  }

  // MARK: - NMSSHChannelDelegate

  public func channel(_ channel: NMSSHChannel, didReadData message: String) {
    guard let connectionId = connections.first(where: { $0.value.session.channel === channel })?.key else { return }
    let base64 = Data(message.utf8).base64EncodedString()
    sendEvent("onData", ["connectionId": connectionId, "base64": base64])
  }

  public func channel(_ channel: NMSSHChannel, didReadError error: String) {
    guard let connectionId = connections.first(where: { $0.value.session.channel === channel })?.key else { return }
    sendEvent("onStatus", ["connectionId": connectionId, "status": "error", "message": error])
  }

  // MARK: - Helpers

  private func requireConfirmed(_ connectionId: String) throws -> LiveConnection {
    guard let conn = connections[connectionId] else { throw VoidSshError.unknownConnection }
    guard conn.hostKeyConfirmed else { throw VoidSshError.hostKeyNotVerified }
    return conn
  }

  private func closeConnection(_ connectionId: String, reason: String) {
    lock.lock()
    let conn = connections.removeValue(forKey: connectionId)
    lock.unlock()
    guard let conn, !conn.closed else { return }
    conn.closed = true
    conn.session.channel.closeShell()
    conn.session.disconnect()
    sendEvent("onStatus", ["connectionId": connectionId, "status": "closed", "message": reason])
  }

  private static func base64NoPad(_ data: Data) -> String {
    var s = data.base64EncodedString()
    while s.hasSuffix("=") { s.removeLast() }
    return s
  }

  private static func writeSshString(_ out: inout [UInt8], _ bytes: [UInt8]) {
    let len = UInt32(bytes.count)
    out.append(UInt8((len >> 24) & 0xFF))
    out.append(UInt8((len >> 16) & 0xFF))
    out.append(UInt8((len >> 8) & 0xFF))
    out.append(UInt8(len & 0xFF))
    out.append(contentsOf: bytes)
  }

  private static func writeSshUInt32(_ out: inout [UInt8], _ value: UInt32) {
    out.append(UInt8((value >> 24) & 0xFF))
    out.append(UInt8((value >> 16) & 0xFF))
    out.append(UInt8((value >> 8) & 0xFF))
    out.append(UInt8(value & 0xFF))
  }

  private static func encodeSshEd25519PublicKey(_ pub: [UInt8]) -> [UInt8] {
    var out: [UInt8] = []
    writeSshString(&out, Array(keyType.utf8))
    writeSshString(&out, pub)
    return out
  }

  /// OpenSSH private-key container (PROTOCOL.key), unencrypted — the file
  /// only ever lives inside the Keychain, never on plain disk. Mirrors
  /// modules/void-ssh/android's Kotlin implementation byte-for-byte; keep
  /// the two in sync if either changes.
  private static func encodeOpenSshEd25519PrivateKey(seed: [UInt8], pub: [UInt8], comment: String) -> String {
    let pubBlob = encodeSshEd25519PublicKey(pub)

    var inner: [UInt8] = []
    var checkintBytes = [UInt8](repeating: 0, count: 4)
    _ = SecRandomCopyBytes(kSecRandomDefault, 4, &checkintBytes)
    inner.append(contentsOf: checkintBytes)
    inner.append(contentsOf: checkintBytes)
    writeSshString(&inner, Array(keyType.utf8))
    writeSshString(&inner, pub)
    writeSshString(&inner, seed + pub) // "sk" field: 32-byte seed || 32-byte pubkey
    writeSshString(&inner, Array(comment.utf8))
    var padByte: UInt8 = 1
    while inner.count % 8 != 0 {
      inner.append(padByte)
      padByte += 1
    }

    var out: [UInt8] = Array("openssh-key-v1".utf8)
    out.append(0) // NUL terminator -- magic is "openssh-key-v1\0", 15 bytes total
    writeSshString(&out, Array("none".utf8)) // ciphername
    writeSshString(&out, Array("none".utf8)) // kdfname
    writeSshString(&out, []) // kdfoptions
    writeSshUInt32(&out, 1) // number of keys
    writeSshString(&out, pubBlob)
    writeSshString(&out, inner)

    let b64 = Data(out).base64EncodedString()
    let wrapped = stride(from: 0, to: b64.count, by: 70).map { i -> String in
      let start = b64.index(b64.startIndex, offsetBy: i)
      let end = b64.index(start, offsetBy: 70, limitedBy: b64.endIndex) ?? b64.endIndex
      return String(b64[start..<end])
    }.joined(separator: "\n")

    return "-----BEGIN OPENSSH PRIVATE KEY-----\n\(wrapped)\n-----END OPENSSH PRIVATE KEY-----\n"
  }
}

File: modules/void-ssh/android/src/main/java/expo/modules/voidssh/VoidSshModule.kt
package expo.modules.voidssh

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import net.schmizz.sshj.SSHClient
import net.schmizz.sshj.common.Buffer
import net.schmizz.sshj.connection.channel.direct.Session
import net.schmizz.sshj.transport.verification.HostKeyVerifier
import net.schmizz.sshj.userauth.keyprovider.OpenSSHKeyFile
import net.schmizz.sshj.userauth.password.PasswordUtils
import org.bouncycastle.crypto.generators.Ed25519KeyPairGenerator
import org.bouncycastle.crypto.params.Ed25519KeyGenerationParameters
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.security.MessageDigest
import java.security.PublicKey
import java.security.SecureRandom
import java.util.Base64
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean

private const val KEY_TYPE = "ssh-ed25519"

private class LiveConnection(val client: SSHClient) {
  @Volatile var hostKeyConfirmed = false
  @Volatile var session: Session? = null
  @Volatile var shell: Session.Shell? = null
  val closed = AtomicBoolean(false)
}

/**
 * Real interactive SSH shell over sshj — see modules/void-ssh/README.md.
 * Written carefully against the sshj 0.38 API but NOT yet compiled or run
 * on a device (no Android SDK in the environment this was authored in).
 * First step on a real machine: `npx expo run:android` and connect to a
 * disposable test server before building anything else on top of this.
 */
class VoidSshModule : Module() {
  private val connections = ConcurrentHashMap<String, LiveConnection>()

  override fun definition() = ModuleDefinition {
    Name("VoidSsh")
    Events("onData", "onStatus")

    // `username` is accepted (not just host/port) so the JS-facing API is
    // identical on both platforms — NMSSH on iOS bakes the username into
    // session construction itself, sshj doesn't need it until auth, but
    // taking it here too means callers never branch on platform.
    AsyncFunction("connect") { host: String, port: Int, _username: String ->
      val id = java.util.UUID.randomUUID().toString()
      val client = SSHClient()
      client.connectTimeout = 12_000

      var fingerprint = ""
      var keyType = ""
      client.addHostKeyVerifier(HostKeyVerifier { _, _, key ->
        keyType = keyTypeName(key)
        fingerprint = sha256Fingerprint(key)
        true // transport-level accept always; app-level trust gate is verifyHostKey() below
      })

      client.connect(host, port)
      connections[id] = LiveConnection(client)
      sendEvent("onStatus", mapOf("connectionId" to id, "status" to "connected"))

      mapOf(
        "connectionId" to id,
        "hostKeyFingerprintSha256" to fingerprint,
        "hostKeyType" to keyType
      )
    }

    AsyncFunction("verifyHostKey") { connectionId: String, accepted: Boolean ->
      val conn = connections[connectionId] ?: throw IllegalStateException("unknown connectionId")
      if (accepted) {
        conn.hostKeyConfirmed = true
      } else {
        closeConnection(connectionId, "host key rejected")
      }
    }

    AsyncFunction("authenticatePassword") { connectionId: String, username: String, password: String ->
      val conn = requireConfirmed(connectionId)
      conn.client.authPassword(username, password)
      sendEvent("onStatus", mapOf("connectionId" to connectionId, "status" to "authenticated"))
    }

    AsyncFunction("authenticateKey") { connectionId: String, username: String, privateKeyOpenSsh: String, passphrase: String? ->
      val conn = requireConfirmed(connectionId)
      val keyFile = OpenSSHKeyFile()
      val finder = if (passphrase != null) PasswordUtils.createOneOff(passphrase.toCharArray()) else null
      keyFile.init(privateKeyOpenSsh.reader(), null, finder)
      conn.client.authPublickey(username, keyFile)
      sendEvent("onStatus", mapOf("connectionId" to connectionId, "status" to "authenticated"))
    }

    AsyncFunction("startShell") { connectionId: String, cols: Int, rows: Int ->
      val conn = connections[connectionId] ?: throw IllegalStateException("unknown connectionId")
      val session = conn.client.startSession()
      session.allocatePTY("xterm-256color", cols, rows, 0, 0, emptyMap())
      val shell = session.startShell()
      conn.session = session
      conn.shell = shell

      val thread = Thread {
        readLoop(connectionId, conn, shell.inputStream)
      }
      thread.isDaemon = true
      thread.start()

      sendEvent("onStatus", mapOf("connectionId" to connectionId, "status" to "shellReady"))
    }

    Function("write") { connectionId: String, data: String ->
      val conn = connections[connectionId] ?: return@Function
      try {
        conn.shell?.outputStream?.let {
          it.write(data.toByteArray(Charsets.UTF_8))
          it.flush()
        }
      } catch (e: Exception) {
        sendEvent("onStatus", mapOf("connectionId" to connectionId, "status" to "error", "message" to (e.message ?: "write failed")))
      }
    }

    Function("resize") { connectionId: String, cols: Int, rows: Int ->
      val conn = connections[connectionId] ?: return@Function
      try {
        conn.session?.let { s ->
          // sshj re-sends a window-change request through the same channel.
          val method = s.javaClass.getMethod("changeWindowDimensions", Int::class.java, Int::class.java, Int::class.java, Int::class.java)
          method.invoke(s, cols, rows, 0, 0)
        }
      } catch (_: Exception) {
        // Best-effort — not all sshj versions expose this the same way; a
        // missed resize just leaves the remote PTY at its previous size.
      }
    }

    Function("disconnect") { connectionId: String ->
      closeConnection(connectionId, "closed by app")
    }

    AsyncFunction("generateEd25519KeyPair") { comment: String ->
      val generator = Ed25519KeyPairGenerator()
      generator.init(Ed25519KeyGenerationParameters(SecureRandom()))
      val pair = generator.generateKeyPair()
      val priv = pair.private as Ed25519PrivateKeyParameters
      val pub = pair.public as Ed25519PublicKeyParameters

      val publicBlob = encodeSshEd25519PublicKey(pub.encoded)
      val privateKeyOpenSsh = encodeOpenSshEd25519PrivateKey(priv.encoded, pub.encoded, comment)
      val fingerprint = "SHA256:" + Base64.getEncoder().withoutPadding().encodeToString(
        MessageDigest.getInstance("SHA-256").digest(publicBlob)
      )
      val publicKeySsh = "$KEY_TYPE ${Base64.getEncoder().encodeToString(publicBlob)} $comment"

      mapOf(
        "privateKeyOpenSsh" to privateKeyOpenSsh,
        "publicKeySsh" to publicKeySsh,
        "fingerprintSha256" to fingerprint
      )
    }

    OnDestroy {
      connections.keys.toList().forEach { closeConnection(it, "module destroyed") }
    }
  }

  private fun requireConfirmed(connectionId: String): LiveConnection {
    val conn = connections[connectionId] ?: throw IllegalStateException("unknown connectionId")
    if (!conn.hostKeyConfirmed) throw IllegalStateException("host key not verified — call verifyHostKey(true) first")
    return conn
  }

  private fun readLoop(connectionId: String, conn: LiveConnection, input: InputStream) {
    val buf = ByteArray(8192)
    try {
      while (!conn.closed.get()) {
        val n = input.read(buf)
        if (n < 0) break
        val chunk = Base64.getEncoder().encodeToString(buf.copyOf(n))
        sendEvent("onData", mapOf("connectionId" to connectionId, "base64" to chunk))
      }
    } catch (e: Exception) {
      if (!conn.closed.get()) {
        sendEvent("onStatus", mapOf("connectionId" to connectionId, "status" to "error", "message" to (e.message ?: "read failed")))
      }
    } finally {
      closeConnection(connectionId, "shell stream ended")
    }
  }

  private fun closeConnection(connectionId: String, reason: String) {
    val conn = connections.remove(connectionId) ?: return
    if (!conn.closed.compareAndSet(false, true)) return
    try { conn.shell?.close() } catch (_: Exception) {}
    try { conn.session?.close() } catch (_: Exception) {}
    try { conn.client.disconnect() } catch (_: Exception) {}
    sendEvent("onStatus", mapOf("connectionId" to connectionId, "status" to "closed", "message" to reason))
  }

  private fun keyTypeName(key: PublicKey): String {
    return when (key.algorithm) {
      "EdDSA", "Ed25519" -> "ssh-ed25519"
      "EC" -> "ecdsa-sha2-nistp256"
      "RSA" -> "ssh-rsa"
      else -> key.algorithm
    }
  }

  /** SHA256 fingerprint over the SSH wire-format key blob (matches `ssh-keygen -lf` output), not the JCA X.509 encoding. */
  private fun sha256Fingerprint(key: PublicKey): String {
    val buf = Buffer.PlainBuffer()
    buf.putPublicKey(key)
    val blob = buf.compactData
    val digest = MessageDigest.getInstance("SHA-256").digest(blob)
    return "SHA256:" + Base64.getEncoder().withoutPadding().encodeToString(digest)
  }

  private fun writeSshString(out: ByteArrayOutputStream, bytes: ByteArray) {
    val len = bytes.size
    out.write((len ushr 24) and 0xFF)
    out.write((len ushr 16) and 0xFF)
    out.write((len ushr 8) and 0xFF)
    out.write(len and 0xFF)
    out.write(bytes)
  }

  private fun writeSshUInt32(out: ByteArrayOutputStream, value: Int) {
    out.write((value ushr 24) and 0xFF)
    out.write((value ushr 16) and 0xFF)
    out.write((value ushr 8) and 0xFF)
    out.write(value and 0xFF)
  }

  private fun encodeSshEd25519PublicKey(pubBytes: ByteArray): ByteArray {
    val out = ByteArrayOutputStream()
    writeSshString(out, KEY_TYPE.toByteArray(Charsets.US_ASCII))
    writeSshString(out, pubBytes)
    return out.toByteArray()
  }

  /**
   * OpenSSH private-key container (PROTOCOL.key), unencrypted
   * ("none"/"none" cipher+kdf) since the file itself is only ever kept
   * inside the Android Keystore-backed Keychain, never on plain disk.
   */
  private fun encodeOpenSshEd25519PrivateKey(seed: ByteArray, pub: ByteArray, comment: String): String {
    val pubBlob = encodeSshEd25519PublicKey(pub)

    val inner = ByteArrayOutputStream()
    val checkint = SecureRandom().nextInt()
    writeSshUInt32(inner, checkint)
    writeSshUInt32(inner, checkint)
    writeSshString(inner, KEY_TYPE.toByteArray(Charsets.US_ASCII))
    writeSshString(inner, pub)
    writeSshString(inner, seed + pub) // "sk" field: 32-byte seed || 32-byte pubkey
    writeSshString(inner, comment.toByteArray(Charsets.UTF_8))
    var padByte = 1
    while (inner.size() % 8 != 0) {
      inner.write(padByte)
      padByte++
    }
    val innerBytes = inner.toByteArray()

    val out = ByteArrayOutputStream()
    out.write("openssh-key-v1".toByteArray(Charsets.US_ASCII))
    out.write(0) // NUL terminator -- magic is "openssh-key-v1\0", 15 bytes total
    writeSshString(out, "none".toByteArray(Charsets.US_ASCII)) // ciphername
    writeSshString(out, "none".toByteArray(Charsets.US_ASCII)) // kdfname
    writeSshString(out, ByteArray(0)) // kdfoptions
    writeSshUInt32(out, 1) // number of keys
    writeSshString(out, pubBlob)
    writeSshString(out, innerBytes)

    val b64 = Base64.getEncoder().encodeToString(out.toByteArray())
    val wrapped = b64.chunked(70).joinToString("\n")
    return "-----BEGIN OPENSSH PRIVATE KEY-----\n$wrapped\n-----END OPENSSH PRIVATE KEY-----\n"
  }
}
